File: androidApp/src/main/kotlin/org/example/project/userInterface/GameViewModel.kt
package org.example.project.userInterface

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import org.example.project.domain.entities.Weapon
import org.example.project.presentation.game.GamePresenter

data class GameUiState(
    val loadedWeapons: List<Weapon> = emptyList(),
    val isLoading: Boolean = false,
    val currentWeapon: Weapon? = null
)

class GameViewModel(
    private val presenter: GamePresenter
) : ViewModel() {

    val uiState: StateFlow<GameUiState> = combine(
        presenter.loadedWeaponsFlow,
        presenter.isLoadingFlow,
        presenter.currentWeaponFlow,
    ) { weapons, isLoading, currentWeapon ->
        GameUiState(
            loadedWeapons = weapons,
            isLoading = isLoading,
            currentWeapon = currentWeapon,
        )
    }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(5000),
        initialValue = GameUiState()
    )

    fun weaponSelected(id: Int) {
        presenter.weaponSelected(id)
    }

    fun resetButtonTapped() {
        presenter.resetButtonTapped()
    }
}

File: iosApp/iosApp/UserInterface/Game/GameViewModel.swift
//
//  GameViewModel.swift
//  iosApp
//
//  Created by ウルトラ深瀬 on 2026/05/31.
//

import Foundation
import Observation
import SharedLogic

@Observable
class GameViewModel {
    private(set) var loadedWeapons: [Weapon] = []
    private(set) var isLoading: Bool = false
    private(set) var currentWeapon: Weapon?
    
    private let presenter: GamePresenter
    private let coroutineCancelHandler: (() -> Void)?
    private var tasks: [Task<Void, Never>] = []
    
    init(
        presenter: GamePresenter,
        coroutineCancelHandler: (() -> Void)?
    ) {
        self.presenter = presenter
        self.coroutineCancelHandler = coroutineCancelHandler
        
        tasks.append(Task { [weak self] in
            for await weapons in presenter.loadedWeaponsFlow {
                self?.loadedWeapons = weapons
            }
        })
        
        tasks.append(Task { [weak self] in
            for await isLoading in presenter.isLoadingFlow {
                self?.isLoading = isLoading.boolValue
            }
        })
        
        tasks.append(Task { [weak self] in
            for await currentWeapon in presenter.currentWeaponFlow {
                self?.currentWeapon = currentWeapon
            }
        })
    }
    
    deinit {
        coroutineCancelHandler?()
        tasks.forEach { task in
            task.cancel()
        }
    }
    
    func weaponSelected(id: Int) {
        presenter.weaponSelected(id: Int32(id))
    }
    
    func resetButtonTapped() {
        presenter.resetButtonTapped()
    }
}
